src/devices/sound/nes_vt_apu.cpp

```diff
@@ -0,0 +1,289 @@
+// license:BSD-3-Clause
+// copyright-holders:David Shah
+/*****************************************************************************
+
+  MAME/MESS VTxx APU CORE
+
+ *****************************************************************************/
+
+#include "nes_vt_apu.h"
+#include "emu.h"
+
+#define VTXX_NTSC_XTAL XTAL(21'477'272)
+#define VTXX_NTSC_APU_CLOCK (VTXX_NTSC_XTAL/12)
+
+DEFINE_DEVICE_TYPE(NES_VT_APU, nesapu_vt_device, "nesapu_vt", "VTxx APU")
+DEFINE_DEVICE_TYPE(NES_VT_APU_SLAVE, nesapu_vt_slave_device, "nesapu_vt_slave", "VTxx APU (slave)")
+
+nesapu_vt_device::nesapu_vt_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+	: nesapu_device(mconfig, tag, NES_VT_APU, owner, clock),
+	  m_xop2(*this, "nesapu_vt_slave"),
+	  m_rom_read_cb(*this)
+{
+
+}
+
+void nesapu_vt_device::device_start()
+{
+	apu_init();
+	m_rom_read_cb.resolve_safe(0x00);
+
+	if(!m_xop2->started())
+		throw device_missing_dependencies();
+	for (int i = 0; i < 2; i++)
+	{
+		save_item(NAME(m_apu_vt.vt33_pcm[i].address), i);
+		save_item(NAME(m_apu_vt.vt33_pcm[i].volume), i);
+		save_item(NAME(m_apu_vt.vt33_pcm[i].enabled), i);
+		save_item(NAME(m_apu_vt.vt33_pcm[i].playing), i);
+		save_item(NAME(m_apu_vt.vt33_pcm[i].phaseacc), i);
+		save_item(NAME(m_apu_vt.vt33_pcm[i].curr), i);
+	}
+
+	save_item(NAME(m_apu_vt.vt03_pcm.phaseacc));
+	save_item(NAME(m_apu_vt.vt03_pcm.regs));
+	save_item(NAME(m_apu_vt.vt03_pcm.address));
+	save_item(NAME(m_apu_vt.vt03_pcm.length));
+	save_item(NAME(m_apu_vt.vt03_pcm.remaining_bytes));
+	save_item(NAME(m_apu_vt.vt03_pcm.enabled));
+	save_item(NAME(m_apu_vt.vt03_pcm.irq_occurred));
+	save_item(NAME(m_apu_vt.vt03_pcm.vol));
+
+	save_item(NAME(m_apu_vt.extra_regs));
+	save_item(NAME(m_apu_vt.vt3x_sel_channel));
+	save_item(NAME(m_apu_vt.use_vt03_pcm));
+	save_item(NAME(m_apu_vt.use_vt3x_pcm));
+
+}
+
+nesapu_vt_slave_device::nesapu_vt_slave_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+	: nesapu_device(mconfig, tag, NES_VT_APU_SLAVE, owner, clock)
+{
+
+}
+
+void nesapu_vt_slave_device::sound_stream_update(sound_stream &stream, stream_sample_t **inputs, stream_sample_t **outputs, int samples)
+{
+	//TODO: any tweaks needed here:
+	nesapu_device::sound_stream_update(stream, inputs, outputs, samples);
+}
+
+void nesapu_vt_device::sound_stream_update(sound_stream &stream, stream_sample_t **inputs, stream_sample_t **outputs, int samples)
+{
+	std::unique_ptr<stream_sample_t[]> pbuf = std::make_unique<stream_sample_t[]>(samples);
+	std::unique_ptr<stream_sample_t[]> sbuf = std::make_unique<stream_sample_t[]>(samples);
+
+	stream_sample_t *pptr = pbuf.get(), *sptr = sbuf.get();
+
+	// Dual legacy sound generators
+	nesapu_device::sound_stream_update(stream, inputs, &pptr, samples);
+	m_xop2->sound_stream_update(stream, inputs, &sptr, samples);
+
+	int accum;
+	memset( outputs[0], 0, samples*sizeof(*outputs[0]) );
```

Comment:
This function isn't consistent in formatting within itself - there are extra spaces inside the parentheses for the `memset` call that aren't used on any of the other function calls here.